test/validation/api/comp/comp.c

```diff
@@ -0,0 +1,558 @@
+ *
+ * SPDX-License-Identifier:	BSD-3-Clause
+ */
+
+#include "config.h"
+
+#include <odp_api.h>
+#include <odp_cunit_common.h>
+#include "test_vectors.h"
+
+#define TEST_NUM_PKT  64
+#define TEST_PKT_LEN  (8 * 1024)
+
+#define SEGMENTED_TEST_PKT_LEN  (16 * 1024)
+#define SEGMENTED_TEST_PATTERN  0xAA
+
+/* Suite names */
+#define ODP_COMP_SYNC_TEST	"Comp/decomp sync test"
+#define ODP_COMP_ASYNC_TEST	"Comp/decomp async test"
+
+#define COMP_PACKET_POOL  "packet_pool"
+#define COMP_OUT_QUEUE	  "comp-out"
+
+
+struct suite_context_s {
+	odp_comp_op_mode_t op_mode;
+	odp_pool_t pool;
+	odp_queue_t queue;
+};
+
+static struct suite_context_s suite_context;
+
+/**
+ * Check if given compression and hash algorithms are supported
+ *
+ * @param comp	Compression algorithm
+ * @param hash	Hash algorithm
+ *
+ * @retval ODP_TEST_ACTIVE when both algorithms are supported
+ * @retval ODP_TEST_INACTIVE when either algorithm is not supported
+ */
+static int check_comp_alg_support(odp_comp_alg_t comp,
+				  odp_comp_hash_alg_t hash)
+{
+	odp_comp_capability_t capability;
+
+	if (odp_comp_capability(&capability))
+		return ODP_TEST_INACTIVE;
+
+	/* Compression algorithms */
+	switch (comp) {
+	case ODP_COMP_ALG_NULL:
+		if (!capability.comp_algos.bit.null)
+			return ODP_TEST_INACTIVE;
+		break;
+	case ODP_COMP_ALG_DEFLATE:
+		if (!capability.comp_algos.bit.deflate)
+			return ODP_TEST_INACTIVE;
+		break;
+	case ODP_COMP_ALG_ZLIB:
+		if (!capability.comp_algos.bit.zlib)
+			return ODP_TEST_INACTIVE;
+		break;
+	case ODP_COMP_ALG_LZS:
+		if (!capability.comp_algos.bit.lzs)
+			return ODP_TEST_INACTIVE;
+		break;
+	default:
+		fprintf(stderr, "Unsupported compression algorithm\n");
+		return ODP_TEST_INACTIVE;
+	}
+
+	/* Hash algorithms */
+	switch (hash) {
+	case ODP_COMP_HASH_ALG_NONE:
+		break;
+	case ODP_COMP_HASH_ALG_SHA1:
+		if (!capability.hash_algos.bit.sha1)
+			return ODP_TEST_INACTIVE;
+		break;
+	case ODP_COMP_HASH_ALG_SHA256:
+		if (!capability.hash_algos.bit.sha256)
+			return ODP_TEST_INACTIVE;
+		break;
+	default:
+		fprintf(stderr, "Unsupported hash algorithm\n");
+		return ODP_TEST_INACTIVE;
+	}
+
+	return ODP_TEST_ACTIVE;
+}
+
+static odp_packet_t run_comp_op(odp_comp_op_t op,
+				odp_comp_alg_t comp_alg,
+				odp_comp_hash_alg_t hash_alg,
+				odp_packet_t inpkt,
+				unsigned int outtext_len)
+{
+	odp_comp_session_t session;
+	odp_comp_capability_t capa;
+	odp_comp_session_param_t ses_params;
+	odp_comp_packet_op_param_t op_params;
+	odp_packet_t outpkt;
+	odp_comp_packet_result_t comp_result;
+	int rc;
+
+	rc = odp_comp_capability(&capa);
+	CU_ASSERT_FATAL(!rc);
+
+	if (comp_alg == ODP_COMP_ALG_NULL &&
+	    !(capa.comp_algos.bit.null))
+		rc = -1;
+	if (comp_alg == ODP_COMP_ALG_DEFLATE &&
+	    !(capa.comp_algos.bit.deflate))
+		rc = -1;
+	if (comp_alg == ODP_COMP_ALG_ZLIB &&
+	    !(capa.comp_algos.bit.zlib))
+		rc = -1;
+	if (comp_alg == ODP_COMP_ALG_LZS &&
+	    !(capa.comp_algos.bit.lzs))
+		rc = -1;
+
+	CU_ASSERT(!rc);
```

Comment:
it's better to put CU_ASSERT in the place where error is actually appers. Error log will have line number which is useful information here.

Reply:
Please compare this with crypto and ipsec code. It all has the same idiom.